File: src/test/java/io/github/aalsanie/codes/IssueOutcomeTest.java
package io.github.aalsanie.codes;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.ArrayList;
import java.util.List;
import org.junit.jupiter.api.Test;

class IssueOutcomeTest {
    @Test
    void createsIssueVariantsWithValueEquality() {
        OutcomeCode code = OutcomeCode.of("com.example.validation", "INVALID_EMAIL");
        assertEquals(Issue.of("Invalid."), Issue.of("Invalid."));
        assertEquals("Invalid.", Issue.of("Invalid.").toString());
        assertEquals("email: Invalid.", Issue.at("email", "Invalid.").toString());
        assertEquals(code + " Invalid.", Issue.coded(code, "Invalid.").toString());
        assertEquals(code + " email: Invalid.", Issue.at("email", code, "Invalid.").toString());
        assertEquals(code, Issue.at("email", code, "Invalid.").getCode());
        assertEquals("email", Issue.at("email", code, "Invalid.").getPath());
        assertNull(Issue.of("Invalid.").getCode());
        assertNull(Issue.of("Invalid.").getPath());
    }

    @Test
    void rejectsInvalidIssueTextAndPath() {
        assertThrows(NullPointerException.class, () -> Issue.of(null));
        assertThrows(IllegalArgumentException.class, () -> Issue.of(" "));
        assertThrows(IllegalArgumentException.class, () -> Issue.of("x".repeat(1025)));
        assertThrows(IllegalArgumentException.class, () -> Issue.of("bad\0message"));
        assertThrows(IllegalArgumentException.class, () -> Issue.at(" ", "Bad."));
        assertThrows(IllegalArgumentException.class, () -> Issue.at("x".repeat(257), "Bad."));
        assertThrows(IllegalArgumentException.class, () -> Issue.at("bad\0path", "Bad."));
        assertThrows(NullPointerException.class, () -> Issue.coded(null, "Bad."));
        assertThrows(NullPointerException.class, () -> Issue.at("field", null, "Bad."));
    }

    @Test
    void outcomeDerivesSemanticsAndDefensivelyCopiesIssues() {
        ArrayList<Issue> source = new ArrayList<>();
        source.add(Issue.at("email", "Invalid."));
        Outcome outcome = Outcome.of(StandardOutcomes.INVALID_ARGUMENT, "request=42", source);
        source.clear();

        assertEquals(StandardOutcomes.INVALID_ARGUMENT, outcome.getDefinition());
        assertEquals(StandardOutcomes.INVALID_ARGUMENT.getCode(), outcome.getCode());
        assertEquals(OutcomeState.FAILED, outcome.getState());
        assertEquals(outcome.getDefaultMessage(), outcome.getMessage());
        assertEquals("request=42", outcome.getDetail());
        assertEquals(1, outcome.getIssues().size());
        assertTrue(outcome.isFailed());
        assertFalse(outcome.isSuccessful());
        assertFalse(outcome.isPending());
        assertTrue(outcome.isTerminal());
        assertFalse(outcome.toString().contains("request=42"));
        assertThrows(UnsupportedOperationException.class, () -> outcome.getIssues().add(Issue.of("x")));
    }

    @Test
    void successPendingAndValidationRulesAreEnforced() {
        Outcome success = Outcome.of(StandardOutcomes.OK);
        assertTrue(success.isSuccessful());
        assertTrue(success.isTerminal());
        assertNull(success.getDetail());

        OutcomeDefinition pendingDefinition = OutcomeDefinition.custom(
            "com.example.jobs", "JOB_PROCESSING", OutcomeState.PENDING, "Job processing."
        );
        Outcome pending = Outcome.of(pendingDefinition, "job=7");
        assertTrue(pending.isPending());
        assertFalse(pending.isTerminal());

        assertThrows(
            IllegalArgumentException.class,
            () -> Outcome.of(StandardOutcomes.OK, null, List.of(Issue.of("No.")))
        );
        assertThrows(IllegalArgumentException.class, () -> Outcome.of(StandardOutcomes.INTERNAL, " "));
        assertThrows(IllegalArgumentException.class, () -> Outcome.of(StandardOutcomes.INTERNAL, "x".repeat(4097)));
        assertThrows(IllegalArgumentException.class, () -> Outcome.of(StandardOutcomes.INTERNAL, "bad\0detail"));
        assertThrows(NullPointerException.class, () -> Outcome.of(null));
        assertThrows(NullPointerException.class, () -> Outcome.of(StandardOutcomes.INTERNAL, null, null));
    }
}

File: src/test/kotlin/io/github/aalsanie/codes/IssueOutcomeTest.kt
package io.github.aalsanie.codes

import org.junit.jupiter.api.Test

class IssueOutcomeTest {
    private val pendingDefinition =
        OutcomeDefinition.custom(
            "com.example.jobs",
            "JOB_PROCESSING",
            OutcomeState.PENDING,
            "The job is processing.",
        )

    @Test fun createsMessageOnlyIssue() {
        val issue = Issue.of("Invalid value.")
        assertNull(issue.code)
        assertNull(issue.path)
        assertEquals("Invalid value.", issue.message)
    }

    @Test fun createsCodedIssue() {
        val code = OutcomeCode.of("com.example", "INVALID_EMAIL")
        val issue = Issue.coded(code, "Invalid email.")
        assertEquals(code, issue.code)
        assertEquals("com.example:INVALID_EMAIL Invalid email.", issue.toString())
    }

    @Test fun createsPathIssue() {
        val issue = Issue.at("user.email", "Invalid email.")
        assertEquals("user.email", issue.path)
        assertEquals("user.email: Invalid email.", issue.toString())
    }

    @Test fun createsPathAndCodeIssue() {
        val code = OutcomeCode.of("com.example", "INVALID_EMAIL")
        val issue = Issue.at("user.email", code, "Invalid email.")
        assertEquals("com.example:INVALID_EMAIL user.email: Invalid email.", issue.toString())
    }

    @Test fun issueHasValueEquality() {
        assertEquals(Issue.at("a", "Bad."), Issue.at("a", "Bad."))
        assertEquals(Issue.at("a", "Bad.").hashCode(), Issue.at("a", "Bad.").hashCode())
        assertNotEquals(Issue.at("a", "Bad."), Issue.at("b", "Bad."))
    }

    @Test fun rejectsBlankIssueMessage() {
        assertFails<IllegalArgumentException> { Issue.of(" ") }
    }

    @Test fun rejectsNulIssueMessage() {
        assertFails<IllegalArgumentException> { Issue.of("bad\u0000message") }
    }

    @Test fun rejectsBlankIssuePath() {
        assertFails<IllegalArgumentException> { Issue.at(" ", "Bad.") }
    }

    @Test fun rejectsOversizedIssuePath() {
        assertFails<IllegalArgumentException> { Issue.at("x".repeat(257), "Bad.") }
    }

    @Test fun rejectsNulIssuePath() {
        assertFails<IllegalArgumentException> { Issue.at("x\u0000y", "Bad.") }
    }

    @Test fun createsSuccessfulOutcome() {
        val outcome = Outcome.of(StandardOutcomes.OK)
        assertTrue(outcome.isSuccessful)
        assertFalse(outcome.isPending)
        assertFalse(outcome.isFailed)
        assertTrue(outcome.isTerminal)
        assertEquals(StandardOutcomes.OK.defaultMessage, outcome.message)
    }

    @Test fun createsPendingOutcomeFromApplicationDefinition() {
        val outcome = Outcome.of(pendingDefinition)
        assertFalse(outcome.isSuccessful)
        assertTrue(outcome.isPending)
        assertFalse(outcome.isFailed)
        assertFalse(outcome.isTerminal)
    }

    @Test fun createsFailedOutcomeWithIssues() {
        val issue = Issue.at("name", "Required.")
        val outcome = Outcome.of(StandardOutcomes.INVALID_ARGUMENT, issues = listOf(issue))
        assertTrue(outcome.isFailed)
        assertEquals(listOf(issue), outcome.issues)
    }

    @Test fun detailDoesNotReplaceStableMessage() {
        val outcome = Outcome.of(StandardOutcomes.NOT_FOUND, detail = "customer 1234 was missing")
        assertEquals(StandardOutcomes.NOT_FOUND.defaultMessage, outcome.message)
        assertEquals("customer 1234 was missing", outcome.detail)
        assertFalse(outcome.toString().contains("customer 1234"))
    }

    @Test fun rejectsIssuesOnSuccessfulOutcome() {
        assertFails<IllegalArgumentException> { Outcome.of(StandardOutcomes.OK, issues = listOf(Issue.of("Bad."))) }
    }

    @Test fun rejectsIssuesOnPendingOutcome() {
        assertFails<IllegalArgumentException> { Outcome.of(pendingDefinition, issues = listOf(Issue.of("Bad."))) }
    }

    @Test fun rejectsBlankDetail() {
        assertFails<IllegalArgumentException> { Outcome.of(StandardOutcomes.OK, " ") }
    }

    @Test fun rejectsOversizedDetail() {
        assertFails<IllegalArgumentException> { Outcome.of(StandardOutcomes.OK, "x".repeat(4097)) }
    }

    @Test fun rejectsNulDetail() {
        assertFails<IllegalArgumentException> { Outcome.of(StandardOutcomes.OK, "x\u0000y") }
    }

    @Test fun outcomeIssuesAreDefensiveCopy() {
        val source = mutableListOf(Issue.of("Bad."))
        val outcome = Outcome.of(StandardOutcomes.INTERNAL, issues = source)
        source.clear()
        assertEquals(1, outcome.issues.size)
    }

    @Test fun outcomeIssuesAreUnmodifiable() {
        val outcome = Outcome.of(StandardOutcomes.INTERNAL, issues = listOf(Issue.of("Bad.")))

        @Suppress("UNCHECKED_CAST")
        val mutable = outcome.issues as MutableList<Issue>
        assertFails<UnsupportedOperationException> { mutable.clear() }
    }
}
